Add Serialize/Deserialize for TaprootSpendInfo

JeremyRubin opened the pull request:
I think this is missing -- unless there is a reason not to have it?

JeremyRubin commented:
builds on #908

apoelstra reviewed (approved):
ACK 2b942cf5066b0e5d614d9a9dc0390ad27ea2f6d2

JeremyRubin commented:
Why?

When #908 is merged, #204f77 will be picked up by git and then this will merge cleanly onto that...

dr-orlovsky commented:
Merging #908 and keeping this one will either require rebase and re-ACK by all reviewers - or will create a duplicated commit in the history.

JeremyRubin commented:
am i missing something about this workflow?

```
❯ git status
On branch master

No commits yet

nothing to commit (create/copy files and use "git add" to track)
❯ touch fi
❯ git add fi
❯ git commit -m 'A'
[master (root-commit) 22c6070] A
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 fi
❯ git checkout -b A
Switched to a new branch 'A'
❯ echo 'Hello' > fi
❯ git add -u
❯ git commit -m "hello"
[A 6ef0ad5] hello
 1 file changed, 1 insertion(+)
❯ git checkout -b B
Switched to a new branch 'B'
❯ echo 'Bye' >> fi
❯ git add -u
❯ git commit -m "bye"
[B 1270c31] bye
 1 file changed, 1 insertion(+)
❯ git checkout master
Switched to branch 'master'
❯ git merge A
Updating 22c6070..6ef0ad5
Fast-forward
 fi | 1 +
 1 file changed, 1 insertion(+)
❯ git merge B
Updating 6ef0ad5..1270c31
Fast-forward
 fi | 1 +
 1 file changed, 1 insertion(+)
❯ git  log --oneline
1270c31 (HEAD -> master, B) bye
6ef0ad5 (A) hello
22c6070 A
```

dr-orlovsky commented:
You are missing the fact that both branches should have the same commit, like in our case

JeremyRubin commented:
```
❯ git log --oneline B
1270c31 (B) bye
6ef0ad5 (A) hello
22c6070 A
(END)

❯ git log --oneline A
6ef0ad5 (A) hello
22c6070 A
(END)
```

JeremyRubin commented:
to be more clear, i've done merges with both #907 and #908 to show the results, either order is fine (altho 1 then 2 creates 1 fewer merge commit).

```
> git checkout -b master_copy
> git merge patch-2
> git merge patch-1
> git log --oneline
afb9d3e (HEAD -> master_copy) Merge branch 'patch-1' into master_copy
680add7 Merge branch 'patch-2' into master_copy
734b1de (origin/master, origin/HEAD) Merge rust-bitcoin/rust-bitcoin#897: Check
for SIGHASH_SINGLE bug in writer fn
2b942cf (fork/patch-1, patch-1) Add Serialize/Deserialize for TaprootSpendInfo
204f477 (fork/patch-2, patch-2) Add serde to TweakedPublicKey
```

```
❯ git checkout -b master_copy2
Switched to a new branch 'master_copy2'
❯ git merge patch-1
Merge made by the 'recursive' strategy.
 src/util/schnorr.rs | 2 ++
 src/util/taproot.rs | 1 +
 2 files changed, 3 insertions(+)
❯ git merge patch-2
Already up to date.
```

apoelstra commented:
@dr-orlovsky you won't have "a duplicated commit in the history". Git will do the right thing when different PRs share a commit, as long as they are literally the same commit (as they are in this case).

dr-orlovsky reviewed (approved):
ACK 2b942cf5066b0e5d614d9a9dc0390ad27ea2f6d2

Sorry for taking your time on this and thank you for the explanations. Let's see if this will work with github: trying to merge this for now

dr-orlovsky commented:
Yep, it worked! I always thought that PRs based on other PRs must be drafts until the underlying PRs got merged... Perfect, thank you!

tcharding commented:
This feature got introduced about  six months to a year ago if I remember correctly. Also (just FYI for your own projects) in some situations you can target the PR that you depend on then only the relevant patches are shown (I think this is when the branch is on the same repo as master i.e., not a PR from personal fork).

apoelstra commented:
Git has supported this for years, as a basic consequence of its DAG model. I'm not sure how Github could've broken that.
